docs/advanced/UsageWithTypescript.md

```diff
@@ -0,0 +1,200 @@
+}
+
+export interface ChatState {
+  messages: Message[]
+}
+```
+
+Describing the state shape of the system reducer's slice of state:
+
+```ts
+// src/store/system/types.ts
+
+export interface SystemState {
+  loggedIn: boolean
+  session: string
+  userName: string
+}
+```
+
+Note that we are exporting these interfaces to reuse them later in reducers and action creators.
+
+## Type Checking Actions & Action Creators
+
+We will be using TypeScript's enums to declare our action constants. [Enums](https://www.typescriptlang.org/docs/handbook/enums.html) allow us to define a set of named constants.
+
+Chat Action Constants & Shape:
+
+```ts
+// src/store/chat/types.ts
+
+export enum ChatActions {
+  SendMessage = 'SEND_MESSAGE'
+}
+
+interface SendMessageAction {
+  type: ChatActions.SendMessage
+  payload: Message
+}
+
+export type ChatActionTypes = SendMessageAction
+```
+
+Note that you can use TypeScript's Union Type to express multiple actions like so:
+
+```ts
+export type ChatActionTypes = SendMessageAction | DeleteMessageAction
+```
+
+With these types declared we can now also type check chat's action creators. In this case we are taking advantage of TypeScript's inference to avoid verbosity:
+
+```ts
+// src/store/chat/actions.ts
+
+import { Message, ChatActions } from './types'
+
+export function sendMessage(newMessage: Message) {
+  return {
+    type: ChatActions.SendMessage,
+    payload: newMessage
+  }
+}
+```
+
+System Action Constants & Shape:
+
+```ts
+// src/store/system/types.ts
+
+export enum SystemActions {
+  UpdateSession = 'UPDATE_SESSION'
+}
+
+interface UpdateSessionAction {
+  type: SystemActions.UpdateSession
+  payload: SystemState
+}
+
+export type SystemActionTypes = UpdateSessionAction
+```
+
+With these types we can now also type check system's action creators:
+
+```ts
+// src/store/system/actions.ts
+
+import { SystemActions, SystemState } from './types'
+
+export function updateSession(newSession: SystemState) {
+  return {
+    type: SystemActions.UpdateSession,
+    payload: newSession
+  }
+}
+```
+
+## Type Checking Reducers
+
+Reducers are just pure functions that take the previous state, an action and then return the next state.
+
+Type checked chat reducer:
+
+```ts
+// src/store/chat/reducers.ts
+
+import { ChatState, ChatActions, ChatActionTypes } from './types'
+
+const initialState: ChatState = {
+  messages: []
+}
+
+export function chatReducer(state = initialState, action: ChatActionTypes) {
+  switch (action.type) {
+    case ChatActions.SendMessage:
+      return {
+        messages: [...state.messages, action.payload]
+      }
+    default:
+      return state
+  }
+}
+```
+
+Type checked system reducer:
+
+```ts
+// src/store/system/reducers.ts
+
+import { SystemActions, SystemState, SystemActionTypes } from './types'
+
+const initialState: SystemState = {
+  loggedIn: false,
+  session: '',
+  userName: ''
+}
+
+export function systemReducer(state = initialState, action: SystemActionTypes) {
```

Comment:
please explicitly indicate the type of return value is `SystemState`, otherwise the return value is not type-safe.

Reply:
```suggestion
export function systemReducer(state = initialState, action: SystemActionTypes): SystemState {
```
it would be better because in practice we may accidentally return a wrong object. this change ensures that you need to always return a `SystemState`.